client.go

```diff
@@ -75,193 +61,242 @@ func (c *Client) Get(ctx context.Context, req *Request) (*GetResult, error) {
+				if err := checksum.Checksum(req.Dst); err == nil {
+					// don't get the file if the checksum of dst is correct
+					getFile = false
+				}
+			}
+			if getFile {
+				if err := g.GetFile(ctx, req); err != nil {
+					os.RemoveAll(td)
+					multierr = append(multierr, err)
+					continue
+				}
+
+				if checksum != nil {
+					if err := checksum.Checksum(req.Dst); err != nil {
+						return nil, err
+					}
+				}
+			}
+
+			if decompressor != nil {
+				// We have a decompressor, so decompress the current destination
+				// into the final destination with the proper mode.
+				err := decompressor.Decompress(decompressDst, req.Dst, decompressDir)
+				if err != nil {
 					return nil, err
 				}
+
+				// Swap the information back
+				req.Dst = decompressDst
+				if decompressDir {
+					req.Mode = ModeAny
+				} else {
+					req.Mode = ModeFile
+				}
+			}
+
+			// We check the dir value again because it can be switched back
+			// if we were unarchiving. If we're still only Get-ing a file, then
+			// we're done.
+			if req.Mode == ModeFile {
+				return &GetResult{req.Dst}, nil
 			}
 		}
 
-		if decompressor != nil {
-			// We have a decompressor, so decompress the current destination
-			// into the final destination with the proper mode.
-			err := decompressor.Decompress(decompressDst, req.Dst, decompressDir)
-			if err != nil {
-				return nil, err
+		// If we're at this point we're either downloading a directory or we've
+		// downloaded and unarchived a directory and we're just checking subdir.
+		// In the case we have a decompressor we don't Get because it was Get
+		// above.
+		if decompressor == nil {
+			// If we're getting a directory, then this is an error. You cannot
+			// checksum a directory. TODO: test
+			if checksum != nil {
+				return nil, fmt.Errorf(
+					"checksum cannot be specified for directory download")
 			}
 
-			// Swap the information back
-			req.Dst = decompressDst
-			if decompressDir {
-				req.Mode = ModeAny
-			} else {
-				req.Mode = ModeFile
+			// We're downloading a directory, which might require a bit more work
+			// if we're specifying a subdir.
+			if err := g.Get(ctx, req); err != nil {
+				os.RemoveAll(td) // Removes temporary folder
+				multierr = append(multierr, err)
+				continue
 			}
 		}
 
-		// We check the dir value again because it can be switched back
-		// if we were unarchiving. If we're still only Get-ing a file, then
-		// we're done.
-		if req.Mode == ModeFile {
-			return &GetResult{req.Dst}, nil
+		// If we have a subdir, copy that over
+		if subDir != "" {
+			if err := os.RemoveAll(realDst); err != nil {
+				return nil, err
+			}
+			if err := os.MkdirAll(realDst, 0755); err != nil {
+				return nil, err
+			}
+
+			// Process any globs
+			subDir, err := SubdirGlob(req.Dst, subDir)
+			if err != nil {
+				return nil, err
+			}
+
+			return &GetResult{realDst}, copyDir(ctx, realDst, subDir, false)
 		}
+		return &GetResult{req.Dst}, nil
```

Comment:
I think it would be better for readability and for this diff if the content of that loop was put into a func:

May be something like:

```go
for _, g := range c.Getters {
	op, err, := c.get(...)...
	// ...
}
```

Reply:
Also it will be easier to see what happens at the error management level etc.

Reply:
Good idea!!